### ISSUE TITLE
Parallel building with meta examples fails

### ISSUE THREAD

**Issue by orivej** [shogun-toolbox/shogun#3999]
I'm building shogun 6.1.2 on Linux, and the parallel building fails because [here](https://github.com/shogun-toolbox/shogun/blob/bc89ca00981a6967310c90c8843a6d19f43c0f11/examples/meta/generator/parse.py#L22) the [generator](https://github.com/shogun-toolbox/shogun/blob/bc89ca00981a6967310c90c8843a6d19f43c0f11/examples/meta/CMakeLists.txt#L68) invokes yacc without explicit `tabmodule` argument, and the default is overwritten by parallel generators.

The error may look like this:
```
[12/872] Generating example meta_api-enums
FAILED: examples/meta/cpp/meta_api/enums.cpp examples/meta/python/meta_api/enums.py examples/meta/octave/meta_api/enums.m examples/meta/java/meta_api/enums.java examples/
meta/ruby/meta_api/enums.rb examples/meta/csharp/meta_api/enums.cs examples/meta/r/meta_api/enums.R examples/meta/lua/meta_api/enums.lua
cd /tmp/nds-build-shogun/source/build/examples/meta && /nix/store/9n2npaz7xy8b4q0a8k28nr3sjzfy625i-python-2.7.14/bin/python /tmp/nds-build-shogun/source/examples/meta/gen
erator/generate.py -i /tmp/nds-build-shogun/source/examples/meta/src/meta_api/enums.sg -o /tmp/nds-build-shogun/source/build/examples/meta -t /tmp/nds-build-shogun/source
/examples/meta/generator/targets -g /tmp/nds-build-shogun/source/build/src/shogun/tags --parser_files_dir=/tmp/nds-build-shogun/source/build/examples/meta/parser_files
Traceback (most recent call last):
  File "/tmp/nds-build-shogun/source/examples/meta/generator/generate.py", line 158, in <module>
    generatedFilesOutputDir=args.parser_files_dir)
  File "/tmp/nds-build-shogun/source/examples/meta/generator/generate.py", line 74, in translateExamples
    ast = parse(file.read(), filePath, generatedFilesOutputDir)
  File "/tmp/nds-build-shogun/source/examples/meta/generator/parse.py", line 320, in parse
    parser = FastParser(generatedFilesOutputDir)
  File "/tmp/nds-build-shogun/source/examples/meta/generator/parse.py", line 25, in __init__
    debug=generateFiles)
  File "/nix/store/fgqqp2033ly1cals1mq27l2f4mi969dm-python2.7-ply-3.8/lib/python2.7/site-packages/ply/yacc.py", line 3264, in yacc
    read_signature = lr.read_table(tabmodule)
  File "/nix/store/fgqqp2033ly1cals1mq27l2f4mi969dm-python2.7-ply-3.8/lib/python2.7/site-packages/ply/yacc.py", line 1967, in read_table
    if parsetab._tabversion != __tabversion__:
AttributeError: 'module' object has no attribute '_tabversion'
```

**Comment by sorig** [comment]
I cannot reproduce this, but it sounds like a race condition caused by PLY caching the generated parser in the `parsetab.py` file on disk. If the parser generator is invoked by multiple processes, this could cause unpredictable behaviour. Potential solutions:

- Disable parser caching in PLY when using parallel builds (is it possible to pass a parameter controlling this from CMake?). This would cause PLY to re-generate the parser for each file being parsed - I don't know how big a performance hit this would be, but it would grow with the number of files we need to parse.
- Maybe we can name the cached parsers based on pid to avoid the processes using the same cache. Downside: a new parser file is generated per process and these are reproduced each time you compile.

**Comment by sorig** [comment]
@orivej Can you elaborate on how you reproduce this exactly? Does it occur just by
```
make -j meta_examples
```
?

**Comment by orivej** [comment]
Yes, `make -j meta_examples` reproduces it every time, but this appears to happen only with a truly fresh build, `rm -rf build; mkdir build; cd build; cmake ...; make -j meta_examples`; `make clean; make -j meta_examples` is not enough.  Sorry if my estimation of the cause is not correct.

**Comment by orivej** [comment]
I have confirmed that the build does not write files outside the `build` directory, so in my case wiping it and unpacking the sources again vs `make clean` may just have an effect on the disk cache or similar that affects timings. I have also recorded with ([fptrace](https://github.com/orivej/fptrace)) that e.g one generator writes
```
build/examples/meta/cpp/binary_classifier/kernel_svm.cpp
build/examples/meta/parser_files/parsetab.py
build/examples/meta/parser_files/parser.out
```
while another writes
```
build/examples/meta/cpp/clustering/gmm.cpp
build/examples/meta/parser_files/parsetab.py
build/examples/meta/parser_files/parser.out
```
so you can tell if this is fixed even without reproduction. Here is one way to fix it: https://github.com/orivej/shogun/commit/11d6bc3150da05fa0eef5451c8e21d3e196b2852

**Comment by vigsterkr** [comment]
@orivej thnx for reporting and writing a possible fix. could you PR this plz?
@sorig are you happy with the fix?

**Comment by sorig** [comment]
I can now reproduce the problem and I've confirmed that @orivej's fix works. Thanks for that!

The problem with this solution is that a new parser is generated for each example file which defeats the point of caching the parser. I've created an alternative solution which generates the parser before the parallel build is run in #4008

**Comment by vigsterkr** [comment]
just merged #4008 that fixes this
